### ISSUE TITLE
Gather layer in TensorRT8.0 get wrong answer when using FP16 mode.

### ISSUE THREAD

**Issue by spxcds** [NVIDIA/TensorRT#1298]
## Description

This is part of my network structure in onnx format.

![截屏2021-06-10 上午9 04 15](https://user-images.githubusercontent.com/9909857/121449247-020f4a80-c9cc-11eb-85d6-7e0cdda91b96.png)

The result is exactly the same when using onnxruntime and tensorrt8.0 with FP32 mode to inference.
But when I set the build config like this

```
config.set_flag(trt.BuilderFlag.FP16)
config.set_flag(trt.BuilderFlag.STRICT_TYPES)
```
Tensor `1657` will get the wrong answer.
I have checked that tensor `1655` works well, and the shape of `1655` is `[1,30,768]`, the type of `1655` is `float`.
The other `Gather` layers which axis equals `0` also works well. 

## Environment

**TensorRT Version**: 8.0EA
**NVIDIA GPU**: T4
**NVIDIA Driver Version**: 440.33.01 
**CUDA Version**: 10.2
**CUDNN Version**: 8.2.0
**Operating System**: centos7
**Python Version (if applicable)**: 3.6
**Tensorflow Version (if applicable)**: 
**PyTorch Version (if applicable)**: 
**Baremetal or Container (if so, version)**:

**Comment by ttyio** [comment]
Hello @spxcds , could you enable verbose log for with and without `fp16 strict` during build TRT engine, and attach the log here? thanks

**Comment by spxcds** [comment]
@ttyio 
This is the log.
[fp16_log.txt](https://github.com/NVIDIA/TensorRT/files/6627957/fp16_log.txt)
I implemented a NonZero plugin in this network. Thanks!

**Comment by ttyio** [comment]
Thanks @spxcds , could you also attach the log without `fp16 strict`, which produce the right result.

Also I have question how did you confirm output `1655` is correct?

**Comment by spxcds** [comment]
@ttyio 
I marked `1655` and `1657` as output separately and compared with onnxruntime's result, so I can confirm that `1655` is correct in `fp16 strict` mode.

This is the log file with `fp16 strict` and `1655` is the output.
[fp16_strict_1655_log.txt](https://github.com/NVIDIA/TensorRT/files/6628241/fp16_strict_1655_log.txt)


And this is the log file without `fp16 strict` and `1657` is the output.
[fp32_1657_log.txt](https://github.com/NVIDIA/TensorRT/files/6628242/fp32_1657_log.txt)

Thanks  a lot!

**Comment by ttyio** [comment]
Hello @spxcds ,
When `1655 ` is marked as output, the previous `1663 + (Unnamed Layer* 74) [Shuffle]...Add_1164` would also run in FP32 precision instead of FP16 because the network output is FP32. 

I cannot repro the `gather`  issue when using a single node in the network. Could you send us a repro? thanks

**Comment by spxcds** [comment]
Hello @ttyio ,
I have send a copy of my model in onnx format and a `libnonzero.so` plugin to your hotmail email.

This is my build code:

```Python
def build():
    ctypes.CDLL('libnonzero.so')
    with trt.Builder(TRT_LOGGER) as builder, \
      builder.create_network(3) as network, \
      builder.create_builder_config() as config, \
      trt.OnnxParser(network, TRT_LOGGER) as parser:
        with open(onnx_model, 'rb') as model:
            if not parser.parse(model.read()):
                for error in range(parser.num_errors):
                    print(parser.get_error(error))
                return None
        profile = builder.create_optimization_profile()

        profile.set_shape('query', (1, 30), (1, 30), (1, 30))
        profile.set_shape('query_mask', (1, 30), (1, 30), (1, 30))

        network.unmark_output(network.get_output(0))

        config.add_optimization_profile(profile)
        config.max_workspace_size = 1 << 32

        config.set_tactic_sources(4)
        # config.set_flag(trt.BuilderFlag.FP16)
        # config.set_flag(trt.BuilderFlag.STRICT_TYPES)

        for i in range(network.num_layers):
            layer = network.get_layer(i)

            if layer.get_output(0).name == output_tensor:  # output_tensor: 1657
                network.mark_output(layer.get_output(0))
                break

        return builder.build_engine(network, config)

```

And this is my input data:

```
query = [101, 4281, 2399, 1920, 100, 4281, 2399, 1920, 102, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
query_mask = [1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
```
Thanks.

**Comment by ttyio** [comment]
Hello @spxcds , if you check the trt layer before the `gather`, it is `1663 + (Unnamed Layer* 74) [Shuffle]...Add_1164` , and if you check your network, there are a lot of nodes fused together. and the difference you see might the FP16 and FP32 difference for this large node.

One experiment to verify is to replace the `gather` with some op like `add(0)`, and dump the tensor after `add(0)`.

**Comment by spxcds** [comment]
> Hello @spxcds , if you check the trt layer before the `gather`, it is `1663 + (Unnamed Layer* 74) [Shuffle]...Add_1164` , and if you check your network, there are a lot of nodes fused together. and the difference you see might the FP16 and FP32 difference for this large node.
> 
> One experiment to verify is to replace the `gather` with some op like `identity` (donot use add because it would fused into the previous large node) `add(0)`, and dump the tensor after `add(0)`.

Thanks,I will try again.

**Comment by ttyio** [comment]
@spxcds 
BTW, if you export your onnx using opset 11 or later, no need to implement `nonzero` plugin, this is an `range` op at begin of your network, thanks

**Comment by spxcds** [comment]
> @spxcds
> BTW, if you export your onnx using opset 11 or later, no need to implement `nonzero` plugin, this seems an `range` op at begin of your network? thanks

Tanks and yes. I will upgrade my export program.
